vm/src/deserializer.h

```diff
@@ -0,0 +1,83 @@
+#ifndef BORGES_DESERIALIZER_H
+#define BORGES_DESERIALIZER_H
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace Borges {
+
+class Deserializer {
+public:
+    Deserializer(std::size_t n, const double* p);
+    void deserialize(double* p);
+    template <typename T>
+    void deserialize(std::shared_ptr<T>* p);
+
+private:
+    std::size_t size_;
+    const double* data_;
+    std::size_t index_;
+    std::unordered_map<std::size_t, std::shared_ptr<void>> index_to_ptr_map_;
+};
+
+void deserialize(Deserializer& d, bool* p);
+void deserialize(Deserializer& d, double* p);
+void deserialize(Deserializer& d, std::size_t* p);
+
+template <typename T>
+T
+deserialize(Deserializer& d)
+{
+    alignas(T) char b[sizeof(T)];
+    T* p = reinterpret_cast<T*>(b);
+    deserialize(d, p);
+    return *p;
+}
+
+template <typename T>
+void
+Deserializer::deserialize(std::shared_ptr<T>* p)
+{
+    std::size_t i = Borges::deserialize<std::size_t>(*this);
+    auto it = index_to_ptr_map_.find(i);
+    if (it != index_to_ptr_map_.end()) {
+        new (p) std::shared_ptr<T>(it->second);
+        return;
+    }
+    std::shared_ptr<T> ptr(std::make_shared<T>(deserialize<T>(*this)));
+    index_to_ptr_map_.insert(std::make_pair(i, std::shared_ptr<void>(ptr)));
+    new (p) std::shared_ptr<T>(std::move(ptr));
```

Comment:
this is magic to me, please explain (in comments)

Reply:
Line 51 constructs a value in the uninitialized storage pointed to by p, using placement new (this is a C++ feature, see http://en.cppreference.com/w/cpp/language/new).

The other lines are simpler. Since this is the first time we've seen this shared_ptr, on line 49 we deserialize the value that it points to, and create a shared_ptr from that (the use of make_shared is an optimization that allocates the refcount and the value itself in a single block). Then on line 50, we associate the index we deserialized earlier with the shared_ptr we just created. If we encounter the same index when deserializing a shared_ptr again, we return the shared_ptr associated with that index instead of deserializing the value again (this is what allows us to preserve shared_ptrs across serialization!).

Reply:
sorry, i made a typo, or the lines have moved, i meant the line

`new (p) std::shared_ptr<T>(std::move(ptr));`

Reply:
That's placement new syntax, which is part of C++03. It works exactly like the new operator you are used to, except that it doesn't allocate memory for the object to be constructed. Instead, it uses the memory pointed to by p to construct the shared_ptr in.

Reply:
ok, i see.